diff --git a/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/Dockerfile b/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/Dockerfile
index 2f40fa271..75200679f 100644
--- a/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/Dockerfile
+++ b/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/Dockerfile
@@ -15,7 +15,6 @@ ARG BASE_IMAGE=rockylinux:8.6
 ARG PYTHON_VERSION=3.9
 ARG GCC_MAJOR_VERSION=12
 ARG IPEX_BRANCH=llm_feature_branch
-ARG PYTORCH_WHL=https://download.pytorch.org/whl/nightly/cpu-cxx11-abi/torch-2.1.0.dev20230715%2Bcpu.cxx11.abi-cp39-cp39-linux_x86_64.whl
 
 FROM ${BASE_IMAGE} AS dev-base
 ARG GCC_MAJOR_VERSION
@@ -37,7 +36,6 @@ ENV PATH /opt/conda/bin:$PATH
 
 FROM dev-base as conda
 ARG PYTHON_VERSION
-ARG PYTORCH_WHL
 RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && \
     chmod +x ~/miniconda.sh && \
     ~/miniconda.sh -b -p /opt/conda && \
@@ -67,9 +65,9 @@ RUN pip install -e git+https://github.com/mlperf/logging@3.0.0-rc2#egg=mlperf-lo
                 onnx==1.14.0 \
                 lark-parser==0.12.0 \
                 hypothesis==6.82.0 \
-                ${PYTORCH_WHL} \
                 pyre-extensions==0.0.30 \
-                scikit-learn==1.3.0
+                scikit-learn==1.3.0 pybind11==2.11.0
+RUN pip install torch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0 --index-url https://download.pytorch.org/whl/cpu
 
 FROM dev-base AS build
 ARG GCC_MAJOR_VERSION
@@ -83,7 +81,7 @@ RUN --mount=type=cache,target=/opt/ccache \
     source /opt/rh/gcc-toolset-${GCC_MAJOR_VERSION}/enable && \
     export CMAKE_PREFIX_PATH=${CONDA_PREFIX:-"$(dirname $(which conda))/../"} && \
     git clone -b ${IPEX_BRANCH} https://github.com/intel/intel-extension-for-pytorch.git intel-extension-for-pytorch && \
-    cd intel-extension-for-pytorch && \
+    cd intel-extension-for-pytorch && git checkout 7256d0848ba81bb802dd33fca0e33049a751db58 && \
     git apply ../ipex.patch && \
     git submodule sync && git submodule update --init --recursive && \
     cd third_party/libxsmm && git checkout c21bc5ddb4 && cd ../../ && \
diff --git a/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/build_dlrm-v2-99_int8_container.sh b/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/build_dlrm-v2-99_int8_container.sh
index 21f9ce4d2..f961585ea 100644
--- a/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/build_dlrm-v2-99_int8_container.sh
+++ b/closed/Intel/code/dlrm-v2-99/pytorch-cpu-int8/docker/build_dlrm-v2-99_int8_container.sh
@@ -9,4 +9,4 @@ export IMAGE_NAME="mlperf_inference_dlrm2:${VERSION}"
 
 echo "Building DLRM2-99.9 workflow container"
 
-DOCKER_BUILDKIT=1 docker build ${DOCKER_BUILD_ARGS} -f Dockerfile -t ${IMAGE_NAME} ../../../..
\ No newline at end of file
+DOCKER_BUILDKIT=1 docker build --network=host ${DOCKER_BUILD_ARGS} -f Dockerfile -t ${IMAGE_NAME} ../../../..
